Readme
/*  Sccsid:     @(#)README	9.1.1.14     1/23/96  09:43:27 */

Table of Contents
===================
 0. What is this document?
 1. What is DBGEN?
 2. What will DBGEN create?
 3. How is DBGEN built?
 4. Command Line Options for DBGEN
 5. Building Large Data Sets with DBGEN
 6. DBGEN limitations and compliant usage
 7. Sample DBGEN executions
 8. What is QGEN?
 9. What will QGEN create?
10. How is QGEN built?
11. Command Line Options for QGEN
12. Query Template Syntax
13. Sample QGEN executions and Query Templates
14. Environment variable
15. Version Numbering in DBGEN and QGEN

0. What is this document?

This is the general README file for DBGEN and QGEN, the data-
base population and executable query text generation programs 
used in the TPC-D benchmark. It covers the proper use of DBGEN and 
QGEN. For information on porting the utility to your particular 
platform see Porting.Notes.

1. What is DBGEN?

DBGEN is a database population program for use with the TPC-D benchmark.
It is written in ANSI 'C' for portability, and has been successfully
ported to over a dozen different systems. While the TPC-D specification
allows an implementor to use any utility to populate the benchmark
database, the resultant population must exactly match the output of
DBGEN. The source code has been provided to make the process of building
a compliant database population as simple as possible.

2. What will DBGEN create?

Without any command line options, DBGEN will generate 8 separate ascii
files. Each file will contain pipe-delimited load data for one of the
tables defined in the TPC-D database schema. The default tables will
contain the load data required for a scale factor 1 database. By 
default the file will be created in the current directory and be 
named <table>.tbl. As an example, customer.tbl will contain the 
load data for the customer table.

When invoked with the '-U' flag, DBGEN will create the data sets to be 
used in the update functions and the SQL syntax required to delete the 
data sets. The update files will be created in the same directory as 
the load data files and will be named "u_<table>.set". The delete 
syntax will be written to "delete.set". For instance, the data set to 
be used in the third query set to update the lineitem table will be 
named "u_lineitem.tbl.3", and the SQL to remove those rows will be 
found in "delete.3". The size of the update files can be controlled 
with the '-r' flag.

3. How is DBGEN built?

Create an appropriate makefile, using makefile.suite as a basis, 
and type make.  Refer to Porting.Notes for more details and for 
suggested compile time options.

4. Command Line Options for DBGEN

DBGEN's output is controlled by a combination of command line options
and environment variables. Command line options are assumed to be single
letter flags preceded by a minus sign. They may be followed by an
optional argument.

option  argument    default     action
------  --------    -------     ------
-v      none                    Verbose. Progress messages are 
                                displayed as data is generated.

-f      none                    Force. Existing data files will be
                                overwritten.

-F      none        yes         Flat file output.

-D      none                    Direct database load. ld_XXXX() routines
                                must be defined in load_stub.c

-s      <scale>     1           Scale of the database population. Scale
                                1.0 represents ~1 GB of data

-T      <table>                 Generate the data for a particular table
                                ONLY. Arguments: p -- part/partuspp, 
                                c -- customer, s -- supplier, 
                                o -- order/lineitem, t -- time, 
                                n -- nation, r -- region,
                                l -- code (same as n and r),
                                O -- order, L -- lineitem, P -- part, 
                                S -- partsupp

-O      d                       Generate SQL for delete function 
                                instead of key ranges

-O      f                       Allow over-ride of default output file 
                                names

-O      h                       Generate headers in flat ascii files.
                                hd_XXX routines must be defined in 
                                load_stub.c

-O      m                       Flat files generate fixed length records

-O      r                       Generate key ranges for the UF2 update 
                                function

-O      s                       Generate the state files for the random
                                number generator.

-O      t                       Generate the optional time table and its
                                associated join fields

-h                              Display a usage summary

-U      <updates>               Create a specified number of data sets
                                in flat files for the update/delete 
                                functions

-r      <percentage>     10     Scale each udpate file to the given 
                                percentage (expressed in basis points)
                                of the data set

-n      <name>                  Use database <name> for in-line load

-C      <children>              Use <children> separate processes to 
                                generate data

-S      <n>                     Generate the <n>th part of a multi-part load

-z                              **generate skewed data distributions**

5. Building Large Data Sets with DBGEN

DBGEN relies on its own random number generator to assure that identical data
sets can be generated on different platforms. In order to build large data sets
using either parallel or multi-stage loads, it is important that the random 
number generator be started, or "seeded", correctly for each step in the load. 
DBGEN includes an option to create correct seed files,  ascii data files used 
to seed the randowm number generator. 

Each line in a seed file represents the state of one part of the random 
number generator after a stage of the data generation has been completed.  
Seed files are named based on the size of the data set being constructed, the
number of children or steps involved in the build, and which step a particular
seed file represents. The naming convention for seed files is mncccsss, where
SF=m * 10^n, ccc is the number of children in the load (in hex) and sss is the
number of the current stage (in hex). For example, the 10th seed file in a 30
process load of 300 GB would be 3201E00A. Since there can be a large number of
seed files, DBGEN allows you to segregate them in the directory named in the
environment variable DSS_SEED.

6. DBGEN limitations and compliant usage

DBGEN is meant to be a robust population generator for use with the 
TPC-D benchmark. It is hoped that DBGEN will make it easier to experi-
ment with and become proficient in the execution of TPC-D's. As a 
result, it includes a number of command line options which are not, 
strictly speaking, necessary to generate a compliant data set for a 
TPC-D run. In addition, some command line options will accept arguments 
which result in the generation of NON-COMPLIANT data sets. Options 
which should be used with care include:

-s -- scale factor. TPC-D runs are only compliant when run against SF's 
      of 1, 10, 30, 100, 300, 1000 ....
-r -- refresh percentage. TPC-D runs are only compliant when run with 
      -r 10, the default.

7. Sample DBGEN executions

DBGEN has been built to allow as much flexibility as possible, but is
fundementally intended to generate two things: a database population 
against which the queries in TPC-D can be run, and the updates that are
used during the update functions in TPC-D. Here are some sample uses
of DBGEN.

  1. To generate the database population for the qualification database
	dbgen -s 0.1
  2. To generate the lineitem table only, for a scale factor 10 database,
     and over-write any existing flat files:
	dbgen -s 10 -f -T L
  3. To build the seed files necessary to load a 30GB data set in 10 steps,
     and include some progress reports:
        dbgen -v -O s -s 30 -C 10
  4. To geterate a 100GB data set in 1GB pieces, generate only the part and 
     partsupplier tables, and include some progress reports along the way:
	dbgen -s 100 -S 1 -C 100 -T p -v (to generate the first 1GB file)
	dbgen -s 100 -S 2 -C 100 -T p -v (to generate the second 1GB file)
        (and so on, incrementing the argument to -S each time)
  5. To generate the update files needed for a 4 stream run of the throughput
     test at 100 GB, using an existing set of seed files from an 8 process 
     load:
	dbgen -s 100 -U 4 -C 8
     Note: since the state of the seed files for a given scale factor is the
     same at the end of the load regardless of the number of children used
     in the load (the same data has been generated, resulting in the same
     modifications to the RNG), the -C argument is arbitrary in the generation
     of updates. Use whatever seed files are available.
     

8. What is QGEN?

QGEN is a query generation program for use with the TPC-D benchmark.
It is written in ANSI 'C' for portability, and has been successfully
ported to over a dozen different systems. While the TPC-D specification
allows an implementor to use any utility to create the benchmark query
sets, QGEN has been provided to make the process of building
a benchmark implementation as simple as possible.

9. What will QGEN create?

QGEN is a filter, triggered by :'s. It does line-at-a-time reads of its
input (more on that later), scanning for :foo, where foo determines the
substitution that occurs. Including:

:<int>          replace with the appropriate value for parameter <int>
:b              replace with START_TRAN (from tpcd.h)
:c              replace with SET_DBASE (from tpcd.h)
:n<int>         replace with SET_ROWCOUNT(<int>) (from tpcd.h)
:o              replace with SET_OUTPUT (from tpcd.h)
:q              replace with query number
:s              replace with stream number
:x              replace with GEN_QUERY_PLAN (from tpcd.h)

Qgen takes an assortment of command line options, controlling which of these
options should be active during the translation from template to EQT, and a
list of query "names". It then translates the template found in
$DSS_QUERY/<name>.sql and puts the result of stdout.

Here is a sample query template:

{  Sccsid:     @(#)1.sql        9.1.1.1     1/25/95  10:51:56  }
:n 0
:o
select
 l_returnflag,
 l_linestatus,
 sum(l_quantity) as sum_qty,
 sum(l_extendedprice) as sum_base_price,
 sum(l_extendedprice * (1 - l_discount)) as sum_disc_price,
 sum(l_extendedprice * (1 - l_discount) * (1 + l_tax)) as sum_charge,
 avg(l_quantity) as avg_qty,
 avg(l_extendedprice) as avg_price,
 avg(l_discount) as avg_disc,
 count(*) as count_order
from lineitem
where l_shipdate <= date '1998-12-01' - interval :1 day
group by l_returnflag, l_linestatus
order by l_returnflag, l_linestatus;

And here is what is generated:
$ qgen -d 1

{return 0 rows}

select
 l_returnflag,
 l_linestatus,
 sum(l_quantity) as sum_qty,
 sum(l_extendedprice) as sum_base_price,
 sum(l_extendedprice * (1 - l_discount)) as sum_disc_price,
 sum(l_extendedprice * (1 - l_discount) * (1 + l_tax)) as sum_charge,
 avg(l_quantity) as avg_qty,
 avg(l_extendedprice) as avg_price,
 avg(l_discount) as avg_disc,
 count(*) as count_order
from lineitem
where l_shipdate <= date('1998-12-01') - interval (90)  day to day
group by l_returnflag, l_linestatus
order by l_returnflag, l_linestatus;

See "Query Template Syntax" below for more detail on converting your prefered query
phrasing for use with QGEN.

10. How is QGEN built?

QGEN is built by the same makefile that creates DBGEN. If the makefile
is successfully creating DBGEN, no further compilation modifications
should be necessary. You may need to modify some of the options which
allow QGEN to integrate with your preferred query tool. Refer to
Porting.Notes for more detail.

11. Command Line Options for QGEN

Like DBGEN, QGEN is controlled by a combination of command line options
and environment variables (See "Environment Variables", below for more
detail).  Command line options are assumed to be single
letter flags preceded by a minus sign. They may be followed by an
optional argument.

option  argument    default     action
------  --------    -------     ------
-c      none                    Retain comments in translation of template to
                                EQT

-d      none                    Default. Use the parameter substitutions
                                required for query validation

-h                              Display a usage summary

-i      <file>                  Use contents of <file> to init a query stream

-l      <file>                  Save query parameters to <file>

-n      <name>                  Use database <name> for queries

-N                              Always use default rowcount, and ignore :n directives

-o      <path>                  Save query n's output in <path>/n.<stream>
                                Uses -p option, and uses :o tag

-p      <stream>                Use the query permutation defined for
                                stream <stream>. If this option is
                                omited, EQT will be generated for the
                                queries named on the command line.

-r      <n>                     Seed the rnadom number generator with <n>

-s      <n>                     Set scale to <n> for parameter 
                                substitutions.

-t      <file>                  Use contents of <file> to complete a query 
                                stream

-T      none                    Use time table format for date substitution

-v      none                    Verbose. Progress messages are 
                                displayed as data is generated.

-x      none                    Generate a query plan as part of query
                                execution.

12. Query Template Syntax

QGEN is a simple ASCII text filter, meant to translate query generalized
query syntax("query template") into the executable query text(EQT) re-
quired by TPC-D. It provides a number of shorthands and syntactic exten-
sions that allow the automatic generation of query parameters and some 
control over the operation of the benchmark implementation.

QGEN first strips all comments from the query template, recognizing both
{comment} and --comment styles. Next it traverses the query template
one line at a time, locating required substitution points, called
parameter tags. The values substituted for a given tag are summarized
below.  QGEN does not support nested substitutions. That is, if
the text substituted for tag itself contains a valid tag the second tag
will not be expanded.

Tag             Converted To            Based on
===             ============            ========
:c		database <dbname>;(1)   -n from the command line
:x              set explain on;(1)      -x from the command line
:<number>       paremeter <number>
:s              stream number
:o              output to outpath/qnum.stream;(1)
					-o from command line, -s from 
                                        command line
:b              BEGIN WORK;(1)          -a from comand line
:e              COMMIT WORK(1)          -a from command line
:q              query number
:n <number>                             sets rowcount to be returned 
                                        to <number>, unless -N appears on the command line

Notes:
   (1)  This is Informix-specific syntax. Refer to Porting.Notes for
   tailoring the generated text to your database environment.
   
13. Sample QGEN executions and Query Templates

QGEN translates generic query templates into valid SQL. In addition, it 
allows conditional inclusion of the commands necessary to connect to a 
database, produce diagnostic output, etc. Here are some sample of QGEN
usage, and the way that command line parameters and the query templates 
interact to produce valid SQL.

  Template, in $DSS_QUERY/1.sql:
            :c
            :o
            select count(*) from foo;
            :x
            select count(*) from lineitem
              where l_orderdate < ':1';

  1. "qgen 1", would produce:
      select count(*) from foo;
      select count(*) from lineitem 
        where l_orderdate < '1997-01-01'; 
   Assuming that 1 January 1997 was a valid substitution for parameter 1.

  2. "qgen -d -c dss1 1, would produce:
      database dss1;
      select count(*) from foo;
      select count(*) from lineitem 
        where l_orderdate < '1995-07-18'; 
   Assuming that 18 July 1995 was the default substitution for parameter 1,
    and using Informix syntax.

  3. "qgen -d -c dss1 -x -o somepath 1, would produce:
      database dss1;
      output to "somepath/1.0"
      select count(*) from foo;
      set explain on;
      select count(*) from lineitem 
        where l_orderdate < '1995-07-18'; 
   Assuming that 18 July 1995 was the default substitution for parameter 1,
    and using Informix syntax.
 

14. Environment Variables

Enviroment variables are used to control features of DBGEN and QGEN 
which are unlikely to change from one execution to another.

Variable    Default     Action
-------     -------     ------
DSS_PATH    .           Directory in which to build flat files
DSS_CONFIG  .           Directory in which to find configuration files
DSS_DIST    dists.dss   Name of distribution definition file
DSS_QUERY   .           Directory in which to find query templates
DSS_SEED    .           Directory in which to find seed files

15. Version Numbering in DBGEN and QGEN

DBGEN and QGEN use a common version numbering algorithm. Each executable
is stamped with a version number which is displayed in the usage messages
available with the '-h' option. A version number is of the form:

   V.R.P.M
   | | | |
   | | | |
   | | | |
   | | |  -- modification: alphabetic, incremented for any trivial changes 
   | | |                   to the source (e.g, porting ifdef's)
   | |  ---- patch level:  numeric, incremented for any minor bug fix
   | |                     (e.g, qgen parameter range)
   | ------- release:      numeric, incremented for each major revision of the
   |                       specification
   |-------- version:      numeric, incremented for each minor revision of the 
                           specification

An implementation of TPC-D is valid only if it conforms to the following
version usage rules:
  -- The Version of DBGEN and QGEN must match the fractional portion of the 
     current specification revision
  -- The Release of DBGEN and QGEN must match the integer portion of the 
     current specification revision
  -- The Modification of DBGEN and QGEN must be the most recent release as
     announced in the minutes from the last General Council meeting of the TPC
  -- The Patch Level must be disclosed as part of the Full Disclosure Report,
     but all unaltered DBGEN/QGEN sources whose version numbers differ only 
     in Patch Level are comparable.

The current revisions are:
  DBGEN: 1.3.1
  QGEN:  1.3.1


History
Changes as of 2/6/97
   -- version 1.3.1
   -- Revised 64 bit support to clean up bcd2_bin() and mk_sparse()
   -- Add 64b support for NT
Changes as of 12/31/97
   -- version 1.3.0
   -- support for seed generation > 1TB (data gen still to be tested)
   -- rework of 64b support
   -- added bcd support for subtraction, comparison, modulo
   -- added 1998 to the copyright notice
   -- clarified comments in dists.dss
   -- corrected substitution problem in Q11
   -- standardized fopen() error messages with OPEN_CHECK()
   -- introduced PATH_SEP in config.h to allow changes in path separators
Changes as of 12/15/96
   -- version 1.2.0
   -- corrected typos in queries 8a, 8c, 8d, 11a, 12F and 14F, 17a
   -- added variant 15c
   -- defined MAX_SCALE and MIN_SCALE; issued error messages for SF > 1000
         since implementation is incomplete
   -- seed file generation can now be resumed with dbgen -R <n> ...
   -- corrected slight compile bug under Solaris 2.5.1
   -- documented compile problems under SunOS
Changes as of 8/1/96
   -- version 1.1.0D
   -- included new variants for queries 8 and 15
   -- re-introduced answer sets in the source tree
Changes as of 5/1/96
   -- version 1.1.0C
   -- unified version numbering of DBGEN and QGEN
   -- updated BUGS list
   -- removed FAQ from soft appendix; web site will keep the current 
           version of the FAQ
   -- added 1996 to the copyright notice
   -- corrected bug in PR_DATE macro; NO CHANGE TO DATA SET
   -- properly initialize param values for cleaner logging
   -- adjusted output format of Q11 partam to allow scaling to 1TB
   -- corrected typos in variant 14c
   -- corrected data type for YEAR in variant 8c
   -- corrected typos in variant 10a
   -- added variant 8d
Changes as of 1/23/96
   -- qgen version 1.1.0B
   -- include support for ANSI semantics
   -- improved patch for seed sensetivity
Changes as of 1/23/96
   -- updated BUGS list
   -- dbgen version 1.1.0A
   -- patch to limit BCD2 fields to 12 characters for columnar output
   -- qgen version 1.1.0A
   -- patch to fix the "unknown flag" problem
   -- patch to fix the seed sensetivity problem
Changes as of 12/19/95
   -- updated BUGS list
   -- dbgen version 1.1.0
   -- upped default value of MAX_CHILDREN to 1000
   -- corrected naming of detail tables in incremental load
   -- corrected range delete output
   -- forced delete files to truncate existing files
   -- removed fixed size tables from seed generation
   -- corrected overflow problem with large scale seed generation
   -- allow date generation as MM-DD-YY based on config.h #define
   -- correct truncation problem with columnar output in PR_VSTR()
   -- added support for Windows NT
   -- added PLATFORM macro to makefile, removed platform defines from
           config.h
   -- removed MAX_CHILDREN define from config.h (set to 1000 in dss.h)
   -- qgen version 1.1.0
   -- correct SET_OUTPUT macro to TDAT
   -- use %ld in output for q17; portability
   -- add support for SQLSERVER database dialect
   -- add support for SYBASE database dialect
   -- adjust parameter ranges for Q1, Q3, Q6
   -- add -T/-t option to usage summary
   -- added support for Windows NT
Changes as of 09/01/95
   -- qgen version 1.0.1
   -- formalized version numbering 
   -- -p now generates correct query permutations
   -- added separate verion number for qgen
   -- corrected Q3 substitution problem
   -- updated permissible range for Q10 
   -- corrected rowcount_dflt and the MAX row indicator (-1)
   -- expanded param logging to include all possible parameters
   -- allowed qgen's -d option to be used at all scale factors
   -- made parameter substitution permutation-independent
   -- added qgen suppport for END_TRAN (-E) and DFLT_NUM (-N)
   -- correct handling of :n directive
   -- added more complete explanation of QGEN to README
   -- rename of random to rndm, for portability
   -- dbgen version 1.0.1
   -- formalized version numbering 
   -- inclusion of SF=1 seed file
   -- correct typo in usage() update example
   -- patch to driver.c to allow correct updates
   -- documentation change to README to clarify seed/stage/update
           intereaction
   -- corrected minor glitch in "open failed" error msg in print.c
   -- added missing line continuation to makefile.suite
   -- seed files are now based on scale factor and number of generators
   -- seed files now hold seeds for one "step" of a given build
   -- clean up of parallel load routines
   -- inclusion of faster seed generation routines from Susanne Englert
   -- removed the -E(xisting) option
   -- assure proper scaling of O_CUSTKEY
   -- corrected default update percentage
   -- proper handling of child tables with '-O f'
   -- removed seed files from the distribution
   -- modified rpb_routine() to limit contribution of partkey in 
           retailprice
   -- added '-S(tep)' option to allow multi-stage loads
   -- roll in of 32 bit speed_seed routines from Dick Shelton
   -- miscelaneous typo corrections in the documentation
   -- cleanup of usage output
Changes as of 05/08/95
   -- version 1.0
   -- add Teradata defines to tpcd.h for QGEN
   -- add :c to query templates for database CONNECT syntax
   -- add examples of DBGEN and QGEN usage to README
   -- add -T option to qgen to allow time table usage
   -- query template names only requre .sql suffix, rest is arbitrary
Changes as of 03/13/95
   -- version 9.1
   -- surround DBNAME with ifndef in config.h
   -- remove -DDBNAME from makefile.suite
   -- sync varchar handling with 9.1 draft
Changes as of 02/21/95
   -- version 9.0a
   -- fixed bug in qgen that incorrectly included rnd.h
   -- included revised DDL with changes for char/varchar and l_quantity
   -- updated DBGEN help message to include new single table options for
           order/lineitem and part/partsupp
   -- included handling for multi-set seed files TPCDSEED.xxx
   -- generated seeds up through 400GB; headed to 1TB!
   -- ANSI lint cleanup; more needed

tpcd.h
/*****************************************************************
 *  Title: tpcd.h for TPC D
 *  Sccsid: @(#)tpcd.h	9.1.1.23 5/1/96  11:45:02
 *  Description:
 *  X
 *
 *****************************************************************
 */
#define DFLT            0x0001
#define OUTPUT          0x0002
#define EXPLAIN         0x0004
#define DBASE           0x0008
#define VERBOSE         0x0010
#define TIMING          0x0020
#define LOG             0x0040
#define QUERY           0x0080
#define REFRESH         0x0100
#define ANSI            0x0200
#define SEED            0x0400
#define COMMENT         0x0800
#define INIT            0x1000
#define TERMINATE       0x2000
#define DFLT_NUM        0x4000

/*
 * general defines
 */
#define VTAG            ':'          /* flags a variable substitution */
#define ofp             stdout       /* make the routine a filter */
#define QDIR_TAG        "DSS_QUERY"  /* variable to point to queries */
#define QDIR_DFLT       "."          /* and its default */

/*
 * database portability defines
 */

//8200 add mysql settings

#ifdef MYSQL
#define GEN_QUERY_PLAN ""
#define START_TRAN "START TRANSACTION"
#define END_TRAN "COMMIT"
#define SET_OUTPUT ""
#define SET_ROWCOUNT "limit %d;\n"
#define SET_DBASE "use %s;\n"
#endif


#ifdef DB2
#define GEN_QUERY_PLAN  "SET CURRENT EXPLAIN SNAPSHOT ON;"
#define START_TRAN      ""
#define END_TRAN        "COMMIT WORK;"
#define SET_OUTPUT      ""
#define SET_ROWCOUNT    "--#SET ROWS_FETCH %d\n"
#define SET_DBASE       "CONNECT TO %s ;\n"
#endif

#ifdef INFORMIX
#define GEN_QUERY_PLAN  "SET EXPLAIN ON;"
#define START_TRAN      "BEGIN WORK;"
#define END_TRAN        "COMMIT WORK;"
#define SET_OUTPUT      "OUTPUT TO "
#define SET_ROWCOUNT    "FIRST %d"
#define SET_DBASE       "database %s ;\n"
#endif

#ifdef 	SQLSERVER
#define GEN_QUERY_PLAN  "set showplan on\nset noexec on\ngo\n"
#define START_TRAN      "begin transaction\ngo\n"
#define END_TRAN        "commit transaction\ngo\n"
#define SET_OUTPUT      ""
#define SET_ROWCOUNT    "set rowcount %d\ngo\n\n"
#define SET_DBASE       "use %s\ngo\n"
#endif

#ifdef 	SYBASE
#define GEN_QUERY_PLAN  "set showplan on\nset noexec on\ngo\n"
#define START_TRAN      "begin transaction\ngo\n"
#define END_TRAN        "commit transaction\ngo\n"
#define SET_OUTPUT      ""
#define SET_ROWCOUNT    "set rowcount %d\ngo\n\n"
#define SET_DBASE       "use %s\ngo\n"
#endif

#ifdef TDAT
#define GEN_QUERY_PLAN  "EXPLAIN"
#define START_TRAN      "BEGIN TRANSACTION"
#define END_TRAN        "END TRANSACTION"
#define SET_OUTPUT      ".SET FORMAT OFF\n.EXPORT REPORT file="
#define SET_ROWCOUNT    ".SET RETCANCEL ON\n.SET RETLIMIT %d\n"
#define SET_DBASE       ".LOGON %s\n"
#endif

#define MAX_VARS      8 /* max number of host vars in any query */
#define QLEN_MAX   2048 /* max length of any query */
#define QUERIES_PER_SET 17
#define MAX_PIDS 50

EXTERN int flags;
EXTERN int s_cnt;
EXTERN char *osuff;
EXTERN int stream;
EXTERN char *lfile;
EXTERN char *ifile;
EXTERN char *tfile;

#define MAX_PERMUTE     100
#ifdef DECLARER
int rowcnt_dflt[QUERIES_PER_SET + 1] = 
    {-1,-1,100,10,-1,-1,-1,-1,-1,-1,20,-1,-1,-1,-1,-1,-1,-1};
int rowcnt;
int permute[MAX_PERMUTE][QUERIES_PER_SET] =
{
{1,4,15,10,11,6,2,16,14,8,12,17,3,5,13,7,9},
{8,2,10,14,12,17,9,3,13,6,15,11,4,1,16,5,7},
{13,15,6,4,12,14,10,1,9,17,3,7,11,5,2,8,16},
{7,3,11,17,1,16,10,9,6,8,4,2,15,5,13,12,14},
{6,11,9,1,14,15,16,3,2,12,17,8,10,7,13,5,4},
{5,7,8,12,13,17,9,3,10,11,6,4,2,1,16,14,15},
{11,9,14,10,17,12,16,2,4,13,3,1,6,5,8,7,15},
{5,4,8,17,13,2,14,9,7,11,3,15,10,16,6,1,12},
{8,3,12,5,10,11,2,4,7,15,1,14,16,17,6,9,13},
{8,4,1,10,3,9,5,16,2,7,6,13,12,11,17,14,15},
{7,10,16,12,4,13,17,3,15,8,9,11,14,6,2,5,1},
{9,16,13,14,15,4,7,8,1,6,11,17,5,10,12,3,2},
{16,13,14,1,9,8,2,12,7,15,11,4,17,10,5,3,6},
{9,4,10,6,12,17,8,11,16,5,2,1,14,13,3,7,15},
{15,14,6,8,13,12,3,11,17,1,2,16,9,5,4,7,10},
{7,5,13,4,15,14,8,3,2,1,9,17,10,16,6,11,12},
{12,4,17,3,6,7,2,8,14,15,16,11,13,10,9,1,5},
{5,7,14,2,16,11,12,13,6,17,3,4,10,9,15,1,8},
{3,11,7,13,10,9,5,8,12,14,15,6,2,1,16,17,4},
{10,15,14,7,5,8,9,12,4,11,17,1,13,3,6,16,2},
{16,1,6,13,11,5,4,15,12,8,14,2,3,10,17,7,9},
{12,15,13,5,14,3,1,9,2,6,16,11,8,4,10,17,7},
{2,11,8,5,9,14,15,16,10,4,6,13,3,1,7,17,12},
{5,1,3,6,15,9,2,12,13,14,11,7,10,17,16,4,8},
{2,11,1,16,3,8,10,13,15,14,9,4,5,6,12,17,7},
{12,3,11,7,16,5,4,9,15,17,2,13,10,6,1,14,8},
{3,5,2,12,1,7,14,11,4,16,10,6,8,9,15,17,13},
{5,8,15,10,14,12,13,16,1,4,3,9,11,17,6,2,7},
{13,3,15,4,1,14,5,10,12,2,11,16,8,6,9,17,7},
{15,2,9,10,6,8,14,4,3,11,17,5,1,7,13,12,16},
{10,2,17,7,9,6,15,12,14,16,11,4,8,3,5,1,13},
{8,5,2,10,12,7,3,6,17,14,16,4,15,9,11,1,13},
{5,16,10,1,6,14,9,4,3,17,7,12,2,11,8,15,13},
{16,17,13,6,15,10,1,2,4,14,3,11,7,8,5,12,9},
{10,5,15,8,4,6,11,12,9,1,13,2,16,14,17,3,7},
{5,11,9,8,16,15,4,2,17,12,7,14,10,13,3,6,1},
{4,15,3,1,8,5,6,12,2,14,17,11,16,13,10,7,9},
{13,7,2,4,6,9,14,3,1,16,17,12,5,11,15,8,10},
{10,16,11,13,8,7,2,9,4,12,14,15,1,6,5,17,3},
{10,12,13,8,3,11,16,1,2,14,6,4,7,5,15,17,9},
{14,9,8,12,13,6,7,10,16,5,3,15,2,17,11,1,4},
{8,12,16,13,15,7,3,2,17,5,11,1,6,14,4,10,9},
{10,3,9,15,8,2,16,4,11,7,6,12,5,1,17,14,13},
{3,16,9,7,15,4,2,1,6,12,11,10,13,14,5,8,17},
{16,17,15,12,3,9,1,13,8,14,5,7,4,11,6,10,2},
{8,12,15,6,3,14,13,17,1,11,5,2,4,9,7,16,10},
{11,16,17,14,10,15,6,2,5,7,12,1,3,4,9,8,13},
{8,12,13,17,6,16,14,5,1,9,7,10,4,15,2,3,11},
{3,2,9,6,15,5,4,7,11,13,12,14,10,8,16,17,1},
{9,3,14,13,7,17,15,12,5,1,8,11,10,16,4,2,6},
{8,7,14,3,6,15,5,2,13,1,10,12,9,11,4,17,16},
{3,13,1,10,8,5,16,14,2,17,12,7,9,6,15,4,11},
{9,7,1,3,6,13,17,11,5,4,10,14,8,12,15,16,2},
{8,1,10,13,16,3,4,14,17,11,12,2,15,5,9,7,6},
{6,14,12,17,16,11,13,9,4,8,10,3,5,2,15,1,7},
{8,17,6,10,11,15,12,3,16,7,1,4,14,2,5,13,9},
{12,13,17,11,6,10,2,1,14,9,7,15,3,5,8,16,4},
{11,2,15,6,9,4,3,12,1,14,10,13,17,7,16,5,8},
{10,5,4,9,2,17,15,6,1,14,3,16,12,11,7,8,13},
{16,10,13,9,8,2,14,5,6,7,17,3,1,15,12,4,11},
{1,3,6,9,16,15,13,11,7,4,8,5,10,14,12,2,17},
{9,14,17,15,16,4,5,3,10,1,13,2,6,11,7,8,12},
{1,3,7,11,16,6,5,9,10,4,15,13,12,8,17,2,14},
{3,16,13,1,15,4,11,7,8,5,9,2,14,10,17,12,6},
{17,3,7,8,6,2,12,15,16,10,14,11,5,4,9,1,13},
{12,15,3,16,5,11,6,10,2,8,4,7,1,13,9,14,17},
{5,4,13,2,16,3,6,15,1,7,14,8,17,9,11,12,10},
{13,1,15,8,2,4,16,11,14,5,12,9,6,3,7,10,17},
{8,1,11,7,17,15,12,5,13,10,14,3,2,4,9,16,6},
{8,3,10,4,13,12,11,1,7,2,9,17,16,15,14,6,5},
{16,9,12,14,17,7,3,6,8,5,11,2,15,1,10,13,4},
{6,9,15,1,7,4,17,8,14,11,13,12,10,16,2,5,3},
{9,10,12,6,17,2,8,7,1,13,16,5,15,14,11,3,4},
{16,4,12,10,5,3,13,11,17,14,15,7,1,2,9,6,8},
{7,14,2,17,6,3,13,12,9,11,10,8,15,5,1,4,16},
{12,7,1,4,14,9,13,16,8,6,10,2,17,11,3,15,5},
{17,10,1,9,3,11,13,4,8,2,12,16,7,15,14,6,5},
{3,17,5,13,12,8,2,4,7,11,15,1,16,14,9,6,10},
{9,8,17,4,7,15,16,13,12,11,3,2,14,1,5,6,10},
{4,14,13,10,6,9,7,2,11,3,16,17,12,1,5,8,15},
{4,2,9,12,6,1,3,17,16,11,7,10,15,14,8,13,5},
{9,15,6,3,11,8,17,1,7,12,14,4,13,16,5,10,2},
{8,11,10,2,9,16,5,6,4,17,14,15,7,1,3,13,12},
{8,1,4,12,3,9,7,10,5,11,13,15,16,2,14,6,17},
{11,9,5,6,10,15,7,1,3,8,12,2,17,4,13,16,14},
{15,4,2,7,10,6,8,13,17,14,11,16,12,9,1,5,3},
{5,8,7,15,14,9,12,10,2,3,17,11,16,4,6,13,1},
{11,16,1,5,14,17,2,9,7,4,6,12,15,10,3,13,8},
{15,13,1,3,6,9,12,14,17,10,16,4,11,5,7,8,2},
{14,8,3,5,11,7,6,16,4,13,9,12,10,17,15,1,2},
{8,11,14,15,13,6,5,3,16,2,12,7,4,9,10,17,1},
{15,1,2,8,5,14,3,12,17,4,10,6,11,9,13,7,16},
{8,12,14,10,1,5,7,16,11,9,4,2,3,13,6,15,17},
{8,13,11,2,10,7,3,12,4,6,14,9,5,15,17,16,1},
{6,12,11,3,10,9,16,1,5,4,13,8,7,2,14,17,15},
{5,6,11,14,10,9,7,17,8,1,16,12,15,3,4,13,2},
{8,1,12,4,13,14,5,11,17,6,7,16,9,3,10,2,15},
{16,13,11,4,5,3,7,6,17,15,1,8,9,14,12,10,2},
{7,12,16,14,13,10,15,8,9,4,6,3,17,2,11,5,1},
{11,6,12,9,8,4,10,1,15,2,14,16,17,13,7,3,5}
};
#define SEQUENCE(stream, query) permute[stream % MAX_PERMUTE][query - 1]
#else
extern int rowcnt_dflt[];
extern int rowcnt;
extern int *permute;
#endif
